osinfo/osinfo_windows.go

```diff
@@ -127,45 +127,40 @@ func getKernelInfo() (string, string, error) {
 	// This should be something like 040904b0 for US English UTF16LE.
 	langCodePage, err := getTranslation(info)
 	if err != nil {
-		return "", "", err
+		return "", "", fmt.Errorf("getTranslation() error: %v", err)
 	}
 
 	return getVersion(info, langCodePage)
 }
 
+type win32OperatingSystem struct {
+	Caption, Version string
+}
+
 // Get reports OSInfo.
 func Get() (*OSInfo, error) {
-	i, err := osInfo()
+	oi := &OSInfo{ShortName: Windows, Architecture: Architecture(runtime.GOARCH)}
+
+	hn, err := os.Hostname()
 	if err != nil {
-		return nil, err
+		return oi, fmt.Errorf("os.Hostname() error: %v", err)
 	}
-
-	oi := &OSInfo{ShortName: Windows, LongName: i.Caption, Version: i.Version, Architecture: Architecture(runtime.GOARCH)}
+	oi.Hostname = hn
 
 	kVersion, kRelease, err := getKernelInfo()
 	if err != nil {
-		return oi, err
+		return oi, fmt.Errorf("getKernelInfo() error: %v", err)
 	}
 	oi.KernelVersion = kVersion
 	oi.KernelRelease = kRelease
 
-	hn, err := os.Hostname()
-	if err != nil {
-		return oi, fmt.Errorf("os.Hostname() error: %v", err)
+	var ops []win32OperatingSystem
+	query := "SELECT Caption, Version FROM Win32_OperatingSystem"
+	if err := wmi.Query(query, &ops); err != nil {
+		return oi, fmt.Errorf("wmi.Query(%q) error: %v", query, err)
 	}
-	oi.Hostname = hn
+	oi.LongName = ops[0].Caption
```

Comment:
Would it be safer to include bounds checking here?

Reply:
Probably a good idea of the calling function changes.